python-package/xgboost/core.py

```diff
@@ -27,12 +28,18 @@
 # lesser tested.  For now we encourage users to pass a simple list of string.
 FeatNamesT = Optional[List[str]]
 
+list_or_dict = TypeVar("list_or_dict", List, Dict)
+ArrayLike = np.ndarray | Sequence
```

Comment:
I think for now we can just use Any.  I have been working on defining the protocol of input data: https://github.com/dmlc/xgboost/pull/7242 We accept a lot more types and ndarray.